### DISCUSSION TITLE
Warning: Refusing to link macOS provided/shadowed software: libffi

### DISCUSSION THREAD

**Discussion by volkerblum** [Homebrew/discussions#1838] [Casks]
Before creating an issue, here is a question that I can't figure out. The background is that I upgraded a Macbook Pro to Big Sur and installed Homebrew from scratch. The (usually difficult) 'grace' installed with only a few manual interventions.

emacs, however, presents a problem. Specifically, it relies on libffi:

    blum:/Users/blum> emacs
    dyld: Library not loaded: /usr/local/opt/libffi/lib/libffi.7.dylib
      Referenced from: /usr/local/opt/p11-kit/lib/libp11-kit.0.dylib
      Reason: image not found
    Abort trap: 6

However:

    blum:/Users/blum> brew link libffi 
    Warning: Refusing to link macOS provided/shadowed software: libffi
    For compilers to find libffi you may need to set:
      export LDFLAGS="-L/usr/local/opt/libffi/lib"
      export CPPFLAGS="-I/usr/local/opt/libffi/include"

    For pkg-config to find libffi you may need to set:
      export PKG_CONFIG_PATH="/usr/local/opt/libffi/lib/pkgconfig"

What to do? emacs appears to be installed correctly:

    blum:/Users/blum> brew info emacs
    emacs: stable 27.2 (bottled), HEAD
    GNU Emacs text editor
    https://www.gnu.org/software/emacs/
    /usr/local/Cellar/emacs/27.2 (4,013 files, 104.3MB) *
      Poured from bottle on 2021-07-16 at 20:17:48
    From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/emacs.rb
    License: GPL-3.0-or-later
    ==> Dependencies
    Build: pkg-config ✔
    Required: gnutls ✔, jansson ✔
    ==> Options
    --HEAD
    	Install HEAD version
    ==> Caveats
    To start emacs:
      brew services start emacs
    Or, if you don't want/need a background service you can just run:
      /usr/local/opt/emacs/bin/emacs --fg-daemon
    ==> Analytics
    install: 26,350 (30 days), 93,516 (90 days), 335,683 (365 days)
    install-on-request: 11,233 (30 days), 41,284 (90 days), 166,625 (365 days)
    build-error: 0 (30 days)

The key point appears to be the refusal to link libffi, claiming (see above):

Warning: Refusing to link macOS provided/shadowed software: libffi

Here, I am lost. I can certainly say

    export DYLD_LIBRARY_PATH='/usr/local/Cellar/libffi/3.3_3/lib/'

and emacs works just fine. However, that seems like a bit of a problematic solution. Any thoughts?

**Comment by gromgit** [comment]
Sounds like your `libffi` install may be br0ken. Try `brew reinstall libffi` and see if that fixes things.

**Comment by volkerblum** [comment]
Thanks! 

Hm: 

    blum:/Users/blum> brew reinstall libffi
    ...
    ==> Reinstalling libffi 
    ==> Pouring libffi--3.3_3.big_sur.bottle.tar.gz
    ==> Caveats
    libffi is keg-only, which means it was not symlinked into /usr/local,
    because macOS already provides this software and installing another version in
    parallel can cause all kinds of trouble.

    For compilers to find libffi you may need to set:
      export LDFLAGS="-L/usr/local/opt/libffi/lib"
      export CPPFLAGS="-I/usr/local/opt/libffi/include"

    For pkg-config to find libffi you may need to set:
      export PKG_CONFIG_PATH="/usr/local/opt/libffi/lib/pkgconfig"

    ==> Summary
    🍺  /usr/local/Cellar/libffi/3.3_3: 17 files, 540.5KB

... seems that someone disabled libffi on purpose. Might be worth filing an issue ... ?

**Reply by gromgit** [reply]
"Keg-only" doesn't mean "disabled". It just means exactly what was stated:
```
it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.
```
Plenty of formulae link with and use libraries from other keg-only formulae just fine.

The reason I suspected your `libffi` install was b0rked is because the "missing" library should exist even in a keg-only install:
```
$ ls -l /usr/local/opt/libffi/lib/libffi.7.dylib
-r--r--r-- 1 aho staff 71112 Jul 17 10:08 /usr/local/opt/libffi/lib/libffi.7.dylib
```

**Reply by volkerblum** [reply]
Yes, sure. But I guess in practice it means disabled since the library cannot be found using the default paths.

First, the term "keg" isn't clear. I understand it is well defined in this community but it will not help users.

Second, it appears that the missing library doesn't exist in my macOS unless one identifies it in Cellar, using the DYLD_LIBRARY_PATH variable. I don't know why. 

Hm ...

Best wishes
Volker

**Reply by gromgit** [reply]
> Yes, sure. But I guess in practice it means disabled since the library cannot be found using the default paths.

Keg-only is _never_ the default choice in Homebrew. It's only used when linking a keg into the default paths is _expected_ to cause problems. That's why all Homebrew formulae link to libraries via `/usr/local/opt` paths, because those always exist regardless of keg-only status.

Which brings us back to your original problem: After `brew reinstall libffi` and _unsetting_ `DYLD_LIBRARY_PATH`, does `emacs` work for you?

**Reply by volkerblum** [reply]
That is a good question. Yes, it actually restored the /usr/local/opt/libffi symlink ... just not any others. But this does work. Thanks!

... now interestingly I found another good question, which is that `brew edit emacs` shows that emacs is compiled without X11 support.

Is that addressable or does the solution lie outside of Homebrew?

(I'm already impressed that grace worked out of the box ... it's such a useful tool)

**Reply by gromgit** [reply]
> Yes, it actually restored the /usr/local/opt/libffi symlink ... just not any others.

Not sure what you mean by "just not any others", but just to be clear:
* `/usr/local/opt` links are created at install time by `brew`
* `brew install <formula>` _only_ takes actions for `<formula>`
* there should be one link for each formula installed (e.g. for every formula `abc` you install, there should be a `/usr/local/opt/abc` link that points to `/usr/local/Cellar/abc/<installed_version>`)
* there should be _no_ link for every formula _not_ installed (i.e. dangling symlinks in `/usr/local/opt` should never exist)

If you notice that a formula you installed does _not_ have a symlink of the same name in `/usr/local/opt`, that's a problem that should be fixed with `brew reinstall <formula>`.

> `brew edit emacs` shows that emacs is compiled without X11 support.
> Is that addressable or does the solution lie outside of Homebrew?

It's addressable, just not within Homebrew core. I don't use Emacs myself, but you have at least two options:
* If you just need "some GUI Emacs" rather than "Emacs specifically for X11", `brew install --cask emacs` should get you a Cocoa-based `Emacs.app`
* A quick online search for "emacs x11 homebrew" surfaces https://github.com/d12frosted/homebrew-emacs-plus

**Reply by carlocab** [reply]
If you need a guide to terminology: https://docs.brew.sh/Formula-Cookbook#homebrew-terminology

That said, I'm not sure why you need it. `brew` has already told you what `keg-only` means:
```
==> Caveats
libffi is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.
```

Which part of this description is confusing?

**Reply by volkerblum** [reply]
Thank you for both answer - much appreciated. Just to complete the thread:

`brew install --cask emacs` does install a version with the right behavior ... however, confusingly, one cannot do so without first uninstalling emacs. That emacs, in turn, is noted as a prerequisite for cask and cannot be uninstalled without overriding. I did override but given that cask was exactly what I was trying to use, that seemed a bit counterintuitive. It did work.

The terminology link is helpful. I agree that the keg-only explanation explained something useful, except that it explained that there would be no symlink under /usr/local/ . I took this to be consistent with my observation that the libffi symlink in /usr/local/opt was missing (which is also "under" /usr/local").

In my case I also was not able to identify exactly where macOS provides libffi.

Additionally, I had actually tried `brew reinstall libffi` before I first posted here, which is confusing. At that point, the symlink under /usr/local/opt was not installed, unlike when I tried again after the third point above.

For me, luckily all this seems to be water under the bridge; the two items that I really needed to install (for now) are there and working. So that's good. I'll now go and find some other libraries I'm missing (actually, I am ... but they are not brew's) in order to begin using this updated OS. 

Thanks again!

**Reply by gromgit** [reply]
> `brew install --cask emacs` does install a version with the right behavior ... however, confusingly, one cannot do so without first uninstalling emacs. That emacs, in turn, is noted as a prerequisite for cask and cannot be uninstalled without overriding. I did override but given that cask was exactly what I was trying to use, that seemed a bit counterintuitive. It did work.

[The `emacs` cask](https://formulae.brew.sh/cask/emacs) declares a _conflict_ with the `emacs` formula, not a dependency on it. In fact, no Homebrew cask declares a dependency on the `emacs` formula, so I'm not sure how you were led to think that the `emacs` cask _and_ formula had to be installed at the same time.

**Reply by volkerblum** [reply]
Hm. I didn't mean the conflict when trying to say brew install --cask emacs. That seemed solvable.

What I thought I would have had to do to resolve that conflict is to `brew uninstall emacs`, the emacs that was installed first. That command declared I could not uninstall emacs unless I chose to manually override ... which I did.

After that, I could `brew install --cask emacs` just fine.

(unfortunately I no longer have the terminal window history associated with this available)

Best wishes again!

**Reply by gromgit** [reply]
> That emacs, in turn, is noted as a prerequisite for cask

Sorry, I just realized you were talking about the `cask` _formula_, which indeed depends on `emacs`. In that situation, you do indeed have to `brew uninstall --ignore-dependencies emacs` to remove the `emacs` formula.
